janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/addedrelations/ConcurrentAddedRelations.java

```diff
@@ -37,11 +41,17 @@ public synchronized boolean remove(final InternalRelation relation) {
 
     @Override
     public synchronized Iterable<InternalRelation> getView(final Predicate<InternalRelation> filter) {
-        return super.getView(filter);
+        return copyView(super.getView(filter));
     }
```

Comment:
I see this method is called from several places, but I can't figure out if there is a bug or not. 
Would you be able to explain why this change is needed? Ideally it would be great to have a test to reproduce the bug, but in some cases it's complicated to write a meaningful test for a bug. In such case a simple description of the problem could be enough.

Reply:
I have added the description of the problem to the PR description.

Reply:
We discussed this part with @ntisseyre previously. 
Current concern is that this method might be called too often which may increase pressure on GC due to creation of many copied arrays in memory.

Below are my observations. 

**Observation 1.**
As I'm checking right now, this method is called for each added Vertex Property. I.e. the below snipped will call this method 3 times:
```
Vertex vertex = tx.addVertex();
vertex.property("foo1", "bar"); // first `getView` call
vertex.property("foo2", "bar2"); // second `getView` call
vertex.property("foo3", "bar3"); // third `getView` call
```

The `addedRelations` (`ConcurrentAddedRelations`) is used in two separate places. One `addedRelations` for the whole transaction and the second `addedRelations` collection for a each vertex. At the beginning I was afraid that it might cause `getView` call to the `addedRelation` of the Transaction which would cause a lot of objects to be copied in-memory (i.e. all relations for the whole transaction). However, upon checking I see that it's not the case. When we are adding `property` that relation is added into both Transaction `addedRelation` collection and StandardVertex `addedRelation` collection, but the `getView` call happens for the StandardVertex `addedRelation` only and not for the whole transaction.
Essentially, it's not a big concern, because properties of a single vertex are copied, and not all properties of all vertices. 

**Observation 2.**
It also seems that when we are executing queries we also call `getNew` method of `StandardJanusGraphTx` which essentially calls `getView` of the Transaction `addedRelation` (i.e. whole transaction). I assume, if you execute many read transaction and also adding many relations in that transaction it might be a problem because with this implementation it will result in many auxiliary arrays to be allocated in-memory to hold all the references. 
That said, looking at `getNew` implementation it seems we already copy those elements into a Set (in one case, but passing a simple reference in another case). And looking at `getUnfoldedIterator()` which is used to process query it seems that the elements returned from `getNew` are copied again into other collections.
As such, the time & space complexity of this flow doesn't actually change. The operations amount change, but amortized complexity stays the same. I.e. if we have N elements in that collection the amount of operations increases from N*2 to N*3 which I believe isn't a big problem. If we want to optimize that then it might be better to optimize N instead of constant.
Overall, I don't think this will cause many problems. The only usage pattern affected I see is when you are adding a huge amount of properties in a single transaction as well as querying some data in parallel for the same transaction. I will assume this isn't the most popular usage pattern.  

Conclusion: I think it's OK to use `copyView` here.

Reply:
thanks!